src/fortran/fdb.f90

```diff
@@ -1,39 +1,329 @@
+      call codes_is_missing(igrib, trim(keyname), is_missing);
+      if (is_missing /= 1) then
+         ! key value is not missing so get values
+         if (present(keyvalue_int)) then
+            call codes_get_int(igrib, keyname, keyvalue_int)
+            write (keyvalue_str, "(I4)") keyvalue_int
+            keyvalue_str=trim(adjustl(keyvalue_str))
+         else
+            call codes_get(igrib, keyname, keyvalue_str)
+         end if
+         write (*, *) keyname, '=', keyvalue_str
+      else
+         write (*, *) keyname, ' is missing from data.'
+      end if
+   END SUBROUTINE
+
+   SUBROUTINE copy_s2a(a,s)   ! copy s(1:Clen(s)) to char array
+      CHARACTER(*),INTENT(IN) :: s
+      CHARACTER :: a(LEN(s)+1)
+      INTEGER :: i
+      DO i = 1,LEN(s)
+         a(i) = s(i:i)
+      END DO
+      a(LEN(s)+1) = char(0)
+   END SUBROUTINE copy_s2a
+
+   SUBROUTINE copy_a2s(a,s)  
+      CHARACTER,INTENT(IN) :: a(:)
+      CHARACTER(len=*),INTENT(INOUT)  :: s
+      INTEGER :: i
+      DO i = 1,SIZE(a)
+         s(i:i) = 'z' ! a(i) ! FAILS HERE
+      END DO
+   END SUBROUTINE copy_a2s
+
+   FUNCTION index_chararray(a,s) result(idx)  
+      CHARACTER,INTENT(IN) :: a(:)
+      CHARACTER(len=*), INTENT(IN)  :: s
+      CHARACTER :: s_a(LEN(s))
+      INTEGER  :: i, j, idx
+      logical :: equal
+      DO i = 1,LEN(s)
+         s_a(i) = s(i:i)
+      END DO
+      DO i = 1,SIZE(a)-SIZE(s_a)
+         DO j = 1, SIZE(s_a)
+            equal = a(i+j-1) == s_a(j)
+            IF (.not. equal) THEN
+               exit
+            END IF
+         END DO
+         IF (equal) THEN
+            idx=i
+            exit
+         END if
+      END DO
+      IF (.not. equal) THEN
+         idx=0
+      END IF
+   END FUNCTION index_chararray
+   
+   SUBROUTINE add_key(key, igrib, keyname_str, type)  
+      use, intrinsic :: iso_c_binding, only : c_int, c_ptr, c_char
+      type(c_ptr), INTENT(INOUT)                  :: key
+      character(len=*), INTENT(IN)                :: keyname_str
+      character(len=*), INTENT(IN), OPTIONAL      :: type
+      ! Local vars
+      character(kind=c_char), dimension(128)      :: keyname
+      character(kind=c_char), dimension(128)      :: keyvalue
+      character(len=128)                          :: keyvalue_str
+      integer                                     :: keyvalue_int
+      integer, INTENT(IN)                         :: igrib
+      integer                                     :: res
+      if (type == 'integer') then
+         call get_value_of_key(igrib, trim(keyname_str), keyvalue_str, keyvalue_int) 
+      else
+         call get_value_of_key(igrib, trim(keyname_str), keyvalue_str) 
+      end if
+      call copy_s2a(keyvalue, trim(keyvalue_str))
+      call copy_s2a(keyname, trim(keyname_str))
+      res = fdb_key_add(key, keyname, keyvalue)
+   END SUBROUTINE add_key
+
+
+   SUBROUTINE convertValues(numStrings, values_str_array, values_array, values_ptr)
```

Comment:
I would add a little description of what the function does

Reply:
added convertValues description: